### ISSUE TITLE
Require `config` in the schema

### ISSUE THREAD

**Issue by wkearn** [ceholden/yatsm#100]
The recent updates to the config schema changed the `reader` section to have a `name` and a `config` section where before it was `name` and `$name` (like `GDAL`). Should `config` be a required entry in the schema?

This just screwed me up temporarily as I was using my old (i.e. Friday's) config file and I got an error about the `input_file` field missing that I had to trace down to this name change, which I just missed as I scanned the new example config. Obviously it's in the new example, so new users who copy that won't hit this. 

It seems that if we're going to throw an error when something in the `config` section (or the section itself) is missing, we should throw it in the config validation step rather than a few steps later when we start trying to load data.

**Comment by ceholden** [comment]
Hey Will

Sorry for moving the earth out from under you, but I think I might have made that change while flying in the air so... anyway in the future I don't think I'll be doing stuff this annoying and/or will do it in a branch.

You found that error at the place you did because the `jsonschema` validation of the configuration file wasn't working correctly. I think the recent commits have fixed this issue. The error message is still scary and probably useless to non-devs, but it's progress that the error is correctly detected and I can wrap this kind of error in the future.

Here's what it looks like when you keep to "GDAL" instead of updating to "config":
``` python
yatsm -v batch HFR1_Tower.yaml 1 1
14:42:00 ERROR yatsm.cli.options options._callback_arg_config:72 Cannot parse config file: "/home/ceholden/Documents/yatsm/examples/HFR1/HFR1_Tower.yaml"
Traceback (most recent call last):
  File "/home/ceholden/Documents/yatsm/yatsm/config/_parse.py", line 63, in validate_config
    return validate_with_defaults(data, schema)
  File "/home/ceholden/Documents/yatsm/yatsm/config/_util.py", line 49, in validate_with_defaults
    DefaultValidatingDraft4Validator(schema).validate(config)
  File "/home/ceholden/conda/envs/yatsm/lib/python3.6/site-packages/jsonschema/validators.py", line 123, in validate
    raise error
jsonschema.exceptions.ValidationError: 'config' is a required property

Failed validating 'required' in schema['properties']['data']['properties']['datasets']['patternProperties']['.*']['properties']['reader']:
    {'properties': {'config': {'oneOf': [{'$ref': '#/definitions/readers/GDAL'},
                                         {'$ref': '#/definitions/readers/AGDCv2'}]},
                    'name': {'enum': ['GDAL', 'AGDCv2']}},
     'required': ['name', 'config']}

On instance['data']['datasets']['Landsat']['reader']:
    {'GDAL': {'band_names': ['blue',
                             'green',
                             'red',
                             'nir',
                             'swir1',
                             'swir2',
                             'temp',
                             'fmask'],
              'column_dtype': {'path': 'str'},
              'date_column': 'datetime',
              'date_format': '%Y-%m-%dT%H:%M:%S.%f',
              'input_file': '/home/ceholden/Documents/landsat_stack/HFR1_Tower//images.csv',
              'keep_open': False},
     'name': 'GDAL'}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ceholden/Documents/yatsm/yatsm/cli/options.py", line 70, in _callback_arg_config
    config = Config.from_file(value)
  File "/home/ceholden/Documents/yatsm/yatsm/api.py", line 52, in from_file
    config = validate_config(parse_config(data))
  File "/home/ceholden/Documents/yatsm/yatsm/config/_parse.py", line 65, in validate_config
    raise InvalidConfigurationException(exc.message)
yatsm.errors.InvalidConfigurationException: 'config' is a required property
Usage: yatsm batch [OPTIONS] CONFIG JOB_NUMBER TOTAL_JOBS

Error: Invalid value for "config": Cannot parse config file /home/ceholden/Documents/yatsm/examples/HFR1/HFR1_Tower.yaml
```

**Comment by ceholden** [comment]
It's probably worth keeping the schema in separate files per component (`pipeline`, `data`, etc) so I can parse/validate them separately, as needed, since this will make it easier to give more specific error messages when things go wrong.

**Comment by wkearn** [comment]
Ahh! Much better. Thanks!

**Comment by ceholden** [comment]
Closed via 63bb7a440749ff612797d9c33c2a659740dea597
